### ISSUE TITLE
Task importance analysis

### ISSUE THREAD

**Issue by ZhugeKongan** [Raschka-research-group/coral-cnn#37]
`def task_importance_weights(label_array):
    uniq = torch.unique(label_array)
    num_examples = label_array.size(0)

    m = torch.zeros(uniq.shape[0])

    for i, t in enumerate(torch.arange(torch.min(uniq), torch.max(uniq))):
        m_k = torch.max(torch.tensor([label_array[label_array > t].size(0), 
                                      num_examples - label_array[label_array > t].size(0)]))
        m[i] = torch.sqrt(m_k.float())

    imp = m/torch.max(m)
    return imp`
IF this code are diffrent from origin paper"Ordinal Regression with Multiple Output CNN for Age Estimation"
![image](https://user-images.githubusercontent.com/49756674/157447574-ca128854-ccd2-44da-984b-cb466daebaf9.png)

**Comment by ZhugeKongan** [comment]
while not ?
`imps=[]
    label = np.array([dataset[i][1] for i in range(len(dataset))])
    # print(label.shape)
    for i in range(max(label)+1):
        imp=len(label[np.where(label==i)])
        imp=np.sqrt(imp*1.0)
        imps.append(imp)
    imps = imps/np.sum(imps)
`

**Comment by ZhugeKongan** [comment]
while not ?
`
imps=[]
    label = np.array([dataset[i][1] for i in range(len(dataset))])
    # print(label.shape)
    for i in range(max(label)+1):
        imp=len(label[np.where(label==i)])
        imp=np.sqrt(imp*1.0)
        imps.append(imp)
    imps = imps/np.sum(imps)
`

**Comment by ZhugeKongan** [comment]
why not?

`def task_importance_weights(dataset):
    imps=[]
    label = np.array([dataset[i][1] for i in range(len(dataset))])
    # print(label.shape)
    for i in range(max(label)+1):
        imp=len(label[np.where(label==i)])
        imp=np.sqrt(imp*1.0)
        imps.append(imp)
    imps = imps/np.sum(imps)
    return torch.from_numpy(np.array(imps))`

**Comment by ZhugeKongan** [comment]
why not ?

    imps=[]
    label = np.array([dataset[i][1] for i in range(len(dataset))])
    # print(label.shape)
    for i in range(max(label)+1):
        imp=len(label[np.where(label==i)])
        imp=np.sqrt(imp*1.0)
        imps.append(imp)
    imps = imps/np.sum(imps)
    return torch.from_numpy(np.array(imps))

**Comment by rasbt** [comment]
Thanks for the comment. Just to make sure we are on the same page, your question is why we used a different task importance weight than Niu et al? We tried the method in Niu et al., and in our case it made the performance worse for both CORAL and the Niu et al. ordinal CNN. So, we tried different methods for task importance and found that the one we included worked best.

**Comment by ZhugeKongan** [comment]
Thanks Reply.  I still have some questions,  the ordinal CNN usually refers to which paper? 
I have the same problem using Niu et al. task importance.  Can i find the origin paper of your code " task_importance_weights(label_array) " ?

Thanks again for your reply!

**Comment by rasbt** [comment]
Ordinal CNN is the method by Niu et al. 2016 you mentioned at the top of this thread ("Ordinal Regression with Multiple Output CNN for Age Estimation").

Sure. We have a task importance discussion in an earlier arxiv version of the paper but decided to remove it from the final paper because it was not super essential.  You can find it here: https://arxiv.org/pdf/1901.07884v4.pdf


<img width="396" alt="Screen Shot 2022-03-09 at 9 15 49 AM" src="https://user-images.githubusercontent.com/5618407/157459336-af228491-f99f-4ce9-82b9-b1233c7b8e31.png">
<img width="365" alt="Screen Shot 2022-03-09 at 9 15 57 AM" src="https://user-images.githubusercontent.com/5618407/157459347-72fc5d21-32de-4623-a803-c3ba39c60bff.png">

**Comment by ZhugeKongan** [comment]
Thanks for reply !
